components/target_estimator.py

```diff
@@ -0,0 +1,102 @@
+from components.chassis import Chassis
+from components.range_finder import RangeFinder
+from components.turret import Turret
+from components.vision import Vision, VisionData
+
+from utilities.functions import constrain_angle
+
+import math
+import time
+
+from typing import Optional
+
+
+class TargetEstimator:
+
+    ALPHA = 0.1  # Exponential smoothing fraction
+    CAMERA_TO_LIDAR = 0.15
+
+    chassis: Chassis
+    range_finder: RangeFinder
+    turret: Turret
+    vision: Vision
+
+    def __init__(self) -> None:
+        self.reset()
+
+    def on_enable(self) -> None:
+        self.reset()
+
+    def reset(self) -> None:
+        self.angle_to_target: float = 0.0
+        self.vision_range: float = 0.0
+        self.previous_azimuth: float = 0.0
+        self.previous_heading: Optional[float] = None
+        self.previous_vision_data: VisionData = VisionData(0.0, 0.0, 0.0)
+        self._pointing_downrange: bool = False
+
+    def is_ready(self) -> bool:
+        return self.is_pointing_downrange() and self.has_vision()
+
+    def has_vision(self) -> bool:
+        return self.vision.is_ready()
+
+    def is_pointing_downrange(self) -> bool:
+        return self._pointing_downrange
+
+    def get_data(self) -> VisionData:
+        if abs(self.angle_to_target) < math.radians(5.0):
+            vd = VisionData(
+                self.range_finder.get_distance() - self.CAMERA_TO_LIDAR,
+                self.angle_to_target,
+                time.monotonic(),
+            )
+        else:
+            vd = VisionData(self.vision_range, self.angle_to_target, time.monotonic(),)
+        return vd
+
+    def execute(self) -> None:
+        # Each timestep we need to adjust our estimate for turret movement.
+        # We also check for new vision data.
+        # If we have it, we add it to our current running average.
+        if not self.is_ready():
+            self.reset()
+        if self.previous_heading is None:
+            self.previous_heading = self.chassis.get_heading()
+
+        # First, correct our estimate
+        current_azimuth = self.turret.get_azimuth()
+        delta = current_azimuth - self.previous_azimuth
+        self.angle_to_target -= delta
+        current_heading = self.chassis.get_heading()
+        delta = constrain_angle(current_heading - self.previous_heading)
+        self.angle_to_target -= delta
+        self.previous_azimuth = current_azimuth
+        self.previous_heading = current_heading
+
+        # Check for new vision data
+        # We only add it if it could possibly be a target
+        # Calculate the turret azimuth in field coordinates
+        turret_in_field = constrain_angle(current_heading + current_azimuth)
+        self._pointing_downrange = abs(turret_in_field) < math.pi / 2.0
```

Comment:
As it stands this relies on the turret azimuth **not** being in robot coordinates. (As far as I can tell, field coordinates are whatever the robot coordinates were when the imu was powered up, i.e. a heading of 0 is pointing up range.) Once the PR to fix that is merged, this will break. The correct test will be turret_in_field < -pi/2 or turret_in_field > pi/2.

Reply:
Field coordinates are fixed at 0 degrees pointing downfield away from our alliance station. So I think that means this is actually correct.

Reply:
So the robot heading starts off at 0 toward the back of the robot, matching the field. Ok. So the turret coordinate system starts off aligned with the field. Cool.

Reply:
And the current azimuth is in robot coordinates (or will be), so the test might not need to change but the computation will, because the field is 180 degrees from robot coordinates.

Reply:
And once the turret get_azimuth is fixed, we'll have to convert it to the field coordinates, so the computation will change.

Reply:
I was under the impression we weren't assuming we were going to have the robot heading start at 0.

Reply:
This code seems to assume that.

Reply:
The autonomous routine knows the heading of the robot and initialises the pose accordingly.